tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlockAwareSegmentInputStreamImpl.java

```diff
@@ -143,6 +191,41 @@ private int readEntries() throws IOException {
         }
     }
 
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        if (len == 0) {
+            return 0;
+        }
+
+        int offset = off;
+        int readLen = len;
+        int readBytes = 0;
+        // reading header
+        if (dataBlockHeaderStream.available() > 0) {
+            int read = dataBlockHeaderStream.read(b, off, len);
+            offset += read;
+            readLen -= readLen;
+            readBytes += read;
+            bytesReadOffset += read;
+        }
+        if (readLen == 0) {
+            return readBytes;
+        }
+
+        // reading ledger entries
+        if (bytesReadOffset < blockSize) {
+            byte[] readEntries = readEntries(readLen);
+            for (int i = 0; i < readEntries.length; i++) {
+                b[offset + i] = readEntries[i];
+                readBytes++;
+            }
+            return readBytes;
+        }
+
+        // reached end
+        return -1;
```

Comment:
If we have read part of bytes from dataBlockHeaderStream and then `bytesReadOffset < blockSize` check failed, is it right to just return `-1` ?

Reply:
Please add a test for this case

Reply:
If the required length is 1024, and the header is 512, it will continue the read the body to fill the rest 512 bytes, right?